Packet++/src/DoIpLayer.cpp

```diff
@@ -0,0 +1,1332 @@
+	{
+		const size_t newPayloadlLength = DOIP_SOURCE_ADDRESS_LEN + DOIP_TARGET_ADDRESS_LEN + data.size();
+		const size_t currentDiagnosticDataLen = m_DataLen - DIAGNOSTIC_DATA_OFFSET;
+
+		setPayloadLength(newPayloadlLength);
+		// always clear the current diagnostic data and extendLayer with the new provided data
+		if (currentDiagnosticDataLen > 0)
+		{
+			shortenLayer(DIAGNOSTIC_DATA_OFFSET, currentDiagnosticDataLen);
+		}
+		extendLayer(DIAGNOSTIC_DATA_OFFSET, data.size());
+		uint8_t* dataPtr = m_Data + DIAGNOSTIC_DATA_OFFSET;
+		memcpy(dataPtr, data.data(), data.size());
+	}
+
+	std::string DoIpDiagnosticMessage::getSummary() const
+	{
+		std::ostringstream oss;
+		oss << "Source Address: " << std::hex << "0x" << getSourceAddress() << "\n";
+		oss << "Target Address: " << std::hex << "0x" << getTargetAddress() << "\n";
+		return oss.str();
+	}
+
+	//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~|
+	// DoIpDiagnosticResponseMessageBase|
+	//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~|
+	DoIpDiagnosticResponseMessageBase::DoIpDiagnosticResponseMessageBase(uint8_t* data, size_t dataLen,
+	                                                                     Layer* prevLayer, Packet* packet)
+	    : DoIpDiagnosticBase(data, dataLen, prevLayer, packet)
+	{}
+
+	DoIpDiagnosticResponseMessageBase::DoIpDiagnosticResponseMessageBase(uint16_t sourceAddress, uint16_t targetAddress,
+	                                                                     DoIpPayloadTypes type)
+	    : DoIpDiagnosticBase(FIXED_LEN)
+	{
+		setHeaderFields(DoIpProtocolVersion::Version02Iso2012, type, (FIXED_LEN - DOIP_HEADER_LEN));
+		setSourceAddress(sourceAddress);
+		setTargetAddress(targetAddress);
+	}
+
+	void DoIpDiagnosticResponseMessageBase::setResponseCode(uint8_t code)
+	{
+		getDiagnosticResponseMessageBase()->diagnosticCode = code;
+	}
+
+	uint8_t DoIpDiagnosticResponseMessageBase::getResponseCode() const
+	{
+		return static_cast<uint8_t>(getDiagnosticResponseMessageBase()->diagnosticCode);
+	}
+
+	const std::vector<uint8_t> DoIpDiagnosticResponseMessageBase::getPreviousMessage() const
+	{
+		if (hasPreviousMessage())
+		{
+			uint8_t* dataPtr = m_Data + PREVIOUS_MSG_OFFSET;
+			return std::vector<uint8_t>(dataPtr, dataPtr + (m_DataLen - PREVIOUS_MSG_OFFSET));
+		}
+		else
+		{
+			return {};
+		}
+	}
+
+	bool DoIpDiagnosticResponseMessageBase::hasPreviousMessage() const
+	{
+		return (m_DataLen > PREVIOUS_MSG_OFFSET);
+	}
+
+	void DoIpDiagnosticResponseMessageBase::setPreviousMessage(const std::vector<uint8_t>& msg)
+	{
+		size_t newPayloadLen = FIXED_LEN - DOIP_HEADER_LEN + msg.size();
+		size_t currentPayloadLen = m_DataLen - PREVIOUS_MSG_OFFSET;
+		setPayloadLength(newPayloadLen);
+		// clear memory for old previous message
+		if (hasPreviousMessage())
+		{
+			shortenLayer(PREVIOUS_MSG_OFFSET, currentPayloadLen);
+		}
+		extendLayer(PREVIOUS_MSG_OFFSET, msg.size());
+		uint8_t* ptr = getDataPtr(PREVIOUS_MSG_OFFSET);
```

Comment:
ditto: instead of shortening and then extending we can calculate the size and decide which action is needed